New Edit File Tool

bassamsdata opened the pull request:
## Description

Add a File Edit tool with deterministic and intelligent matching strategies.

### Overview

Introduces the renewed `insert_edit_into_file`, a tool that allows LLMs to make reliable, deterministic file edits using structured function calls. It supports substring and block edits, full overwrites, and boundary insertions while maintaining atomic behavior, all edits succeed or none are applied.

### Key Features

* **Atomic operations** – applies all edits or none to prevent partial file corruption
* **Multiple matching strategies** – exact, whitespace-normalized, punctuation-tolerant, block-anchored, and substring-based
* **Substring mode** – efficient for short token replacements (max 1000 occurrences)
* **Boundary markers** – insert at file start (`^` or `<<START>>`) or end (`$` or `<<END>>`)
* **Complete overwrite mode** – replace entire file content efficiently
* **Error diagnostics** – context-aware messages with similarity suggestions and line-level feedback
* **Conflict detection** – prevents overlapping or incompatible edits
* **File limits** – up to 2MB files and 50KB search text per edit

## LLM Interface

### Complete JSON Schema

```json
{
  "filepath": "src/utils.js",
  "mode": "append",
  "dryRun": false,
  "explanation": "Refactor function names and improve logging",
  "edits": [
    {
      "oldText": "function calc(a, b) {\n  return a + b;\n}",
      "newText": "function calculate(first, second) {\n  return first + second;\n}"
    },
    {
      "oldText": "console.log('debug')",
      "newText": "logger.info('debug')",
      "replaceAll": true
    }
  ]
}
```

### Parameter Reference

Required:

* `filepath` – path to the target file
* `edits` – array of edit operations

Optional:

* `mode` – `"append"` (default) or `"overwrite"` (replace file)
* `dryRun` – preview changes before writing (`false` by default)
* `explanation` – brief text describing the edit

Per-edit options:

* `oldText` – text to find
* `newText` – replacement text
* `replaceAll` – replace all occurrences (`false` by default)

### Special Cases

Overwrite mode:

```json
{
  "filepath": "config.js",
  "mode": "overwrite",
  "edits": [
    {"oldText": "", "newText": "const config = { api: 'v2', debug: false };"}
  ]
}
```

File boundaries:

```json
{
  "filepath": "script.py",
  "edits": [
    {"oldText": "^", "newText": "#!/usr/bin/env python3\n"},
    {"oldText": "$", "newText": "\nprint('Done')"}
  ]
}
```

## Technical Implementation

### Architecture Overview

1. **Entry Point** – separates substring and block edits, ensuring atomic application
2. **Sequential Processing** – substring edits handled in parallel, block edits sequentially
3. **Matching Chain** – tries multiple fallback strategies with confidence scoring:

   * exact_match (1.0)
   * substring_exact_match (1.0)
   * whitespace_normalized (0.95)
   * punctuation_normalized (0.93)
   * position_markers (1.0)
   * trimmed_lines (0.8)
   * block_anchor (0.6)
4. **Conflict Detection** – identifies overlapping or conflicting regions
5. **Error Reporting** – suggests near-matches, shows line numbers, and hints for fixable cases

### Testing Summary

Validated with over 20 LLMs, handling up to 100 edits per file and 1000-line test cases. Tested providers and models include:

* **Proprietary Models:** OpenAI Models including the old 4.1, Gemini, Grok XAI
* **Open Source Models:** Qwen 3 Coder, Kimi K2, Gym 4.6, Gym 4.5 Air, Owen 14B, Owen 30b and OpenAI OSS series

The tool correctly handled structured JSON, quoted JSON, and Python-like boolean syntax across all tested models.

### Status

✅ Stable and ready for general use.



## Related Issue(s)

[https://github.com/olimorris/codecompanion.nvim/discussions/1607#discussioncomment-14397658](https://github.com/olimorris/codecompanion.nvim/discussions/1607#discussioncomment-14397658)
#2128
#2029
#1619



## Screenshots

<!-- Add screenshots of the changes if applicable, to help visualize the change. -->

## Checklist

- [x] I've read the [contributing](https://github.com/olimorris/codecompanion.nvim/blob/main/CONTRIBUTING.md) guidelines and have adhered to them in this PR
- [x] I've added [test](https://github.com/olimorris/codecompanion.nvim/blob/main/CONTRIBUTING.md#testing) coverage for this fix/feature
- [x] I've run `make all` to ensure docs are generated, tests pass and my formatting is applied
- [ ] _(optional)_ I've updated `CodeCompanion.has` in the [init.lua](https://github.com/olimorris/codecompanion.nvim/blob/main/lua/codecompanion/init.lua#L239) file for my new feature
- [ ] _(optional)_ I've updated the README and/or relevant docs pages

olimorris commented on lua/codecompanion/config.lua:
```diff
@@ -64,12 +64,13 @@ local defaults = {
               "get_changed_files",
               "grep_search",
-              "insert_edit_into_file",
+              -- "insert_edit_into_file",
```
This should replace the `insert_edit_into_file` tool. So when this is ready, I propose we rename it to that

olimorris commented on lua/codecompanion/strategies/chat/tools/catalog/edit_tool_exp.lua:
```diff
@@ -0,0 +1,1022 @@
+    type = "function",
+    ["function"] = {
+      name = "edit_tool_exp",
+      description = "Robustly edit files with multiple automatic fallback strategies. Handles whitespace differences, indentation variations, and ambiguous matches intelligently.",
```
Could we move the system prompt into the tool's description? I'm hoping to piece together an "agent" like mode very soon and really want to avoid tool specific usage in the system prompt.

olimorris commented on tests/strategies/chat/tools/catalog/test_edit_tool_exp.lua:
```diff
@@ -0,0 +1,1028 @@
+T["Basic Functionality"]["can edit a simple file"] = function()
+  child.lua([[
+    -- create initial file
+    local initial = "function getName() {\n  return 'John';\n}"
```
I propose that we put this "initial" or "old_text" in a stub file. It will make it easier in the future to test specific code snippets for specific filetypes

Could name them:

- `simple_edit.php.oldText`
- `simple_edit.php.newText`

and create a helper function like `create_stub("simple_edit.php")` which returns the contents of both.

I'm happy to own this btw.

bassamsdata replied:
This is much better for future testing.

> I'm happy to own this btw.

Honestly, that would be fantastic if you have the time, thank you. If you don’t, no worries; I’ll take care of it.

olimorris replied:
Of course. It feels like a sensible paradigm to adopt for testing in the plugin anyway

olimorris reviewed (commented):
Completely onboard with this approach. The strategy chain is a brilliant idea alongside the confidence levels. Also, your testing is as comprehensive as it could get. This will be an amazing feature.

Sorry if I made you come back to the laptop by carrying on that conversation. No pressure at all.

bassamsdata commented:
Hi @olimorris, hope you're doing well

Apologies for the delay in updating this PR. I’ve cleaned and organized the tool today, and it's nearly ready, I just need to do a bit more prompt testing for the new refactored prompt.
Before I mark it as "ready for review" I wanted to check in with you on the naming. Would you still prefer to stick with `insert_edit_into_file`, or would `edit_file` be a better fit? Personally, I lean toward `edit_file` since it more clearly reflects what the tool does.
I understand this might be a breaking change, so I’ll leave it up to you, I'm happy to update it either way, right away.

Thanks

olimorris commented:
Please don't feel any pressure regarding any PR for CodeCompanion. Time is always going to be our enemy and we should prioritize life before this plugin 😄.

> Would you still prefer to stick with insert_edit_into_file

Yes. I want to continue to try and have parity between Copilot and CodeCompanion regarding tools, where possible. Also, it's one less thing to notify and potentially annoy some users with.

bassamsdata commented:
Hey @olimorris,

This is ready for review! Functionality-wise, I’m happy with how it’s working. There might be a few small prompt tweaks needed later on, but only if any issues come up.

There’s a minor issue with the winbar I just noticed, that I’ll try to sort out while you’re reviewing.

One quick note on the code, there are a few strategies we might eventually merge or even delete, but I think it’s best to keep them for now so we can gather some initial user feedback first.

Thanks

olimorris commented on lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua:
```diff
@@ -0,0 +1,972 @@
+    ---@param tools CodeCompanion.Tools
+    ---@param config table The tool configuration
+    ---@return boolean
+    prompt_condition = function(self, tools, config)
```
We're overwriting the `config` at the top of the file. This was probably a legacy issue from my previous implementation

olimorris commented on lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua:
```diff
@@ -0,0 +1,972 @@
+-- Enhanced Python-like parser for handling LLM's mixed Python/JSON syntax
+---@param edits_string string The string containing edits in Python-like format
+---@return string The converted JSON-like string
+local function parse_python_like_edits(edits_string)
```
`edits` is better than `edits_string` as we state the type is a string in the doc block above

olimorris commented on lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua:
The review point I missed...My preference is to have internal filepath parameters as `path` as it's just a more common Neovim/Vim term. However, externally facing (i.e. when talking to an LLM, keeping it as filepath. So things like `action.filepath` are totally fine but in `write_file_content` we should have that parameter as `path`.

I actually used the `insert_edit_into_file` tool to make those changes and it nailed it first time:

<img width="3368" height="2232" alt="Image" src="https://github.com/user-attachments/assets/a3290bfc-71fc-484a-b64e-768681dc40d6" />

olimorris commented:
The other point I wanted to add was regarding the output messages in the chat buffer. With the last edit I made, I see:

```markdown
Applied 8 edit(s) to lua/codecompanion/strategies/chat/tools/catalog/insert_edit_into_file/init.lua
Rename all internal function parameters and variables from `filepath` to `path` while preserving `filepath` in the schema and user-facing error messages
```

For brevity, it would be cleaner if that first line read `Edited insert_edit_into_file/init.lua file` which, I recall is how Copilot Chat does it. As the user will likely have a diff to manually view the edits then the "applied" and "8 edits" is superfluous, saving us some space in the chat buffer. The reasoning message, however, is superb and it's great to see that.

bassamsdata commented:
Sure, I’ll update that first line to say `Edited insert_edit_into_file/init.lua file`.

I’ll also tweak the `fmt()` output to clean up the English style (so we don’t get things like `1 edit(s)`). Adding this helper should do the trick in the dryRun ofcourse.

```lua
--- Return a word in singular or plural form based on count
function M.pluralize(count, word)
  return count == 1 and word or word .. "s"
end
```

olimorris reviewed (commented):
This is amazing, @bassamsdata. Just super comprehensive, well thought out and I've not been able to break it 😆.

Comments below are super minor. I made them over the course of 12+ hours so they might not be very coherent so let me know if they don't make sense.

olimorris commented:
Firstly, an apology. I didn't submit my preview on Sunday. So these comments have been sat here for a whole week as "pending". I'm an idiot.

My feedback is mostly minor, I did observe some LuaLs warnings but otherwise it's just nit picky stuff.

bassamsdata commented:
No worries at all! Most of the review comments were already addressed during my refactoring, and I’ve just resolved the remaining ones. I also added an option for the file size:

```lua
file_size_limit_mb = 2, -- Maximum file size in MB
```

Though worth mentioning, I also added two new files after your review, one for constants and another for `text_utils`, which I extracted from the strategies to make the code more maintainable.

Thanks again for the feedback!

olimorris commented:
One final thought. Is it worth including a sentence or two for a system prompt?

I added a memory tool last night. It's native to Anthropic so their models had no issue using it. But I had to prompt OpenAI quite heavily to use it. Only when I included a lite system prompt did they start using it like Anthropic.

bassamsdata commented:
> One final thought. Is it worth including a sentence or two for a system prompt?

Done, I hope it’s not too much, but please let me know if it feels like too much.

> I added a memory tool last night.

Nice addition! When I rebased last night, I noticed it, though my late-night brain briefly thought it was the other kind of memory and wondered _'hmm, memory in the tools section!'_ 😅

I’m happy with the current state of the PR. If you’re good with it too, feel free to merge!

olimorris commented:
Amazing. Thanks so much. I'll write the announcement now.
